### ISSUE TITLE
Deadlock / buffer full(?) in mamba output

### ISSUE THREAD

**Issue by xhochy** [QuantStack/mamba#255]
I sometimes get a deadlock with the output being truncated like this. Happens like 5% of all installs, where 90% of the 5% are inside a Github Actions job. I'm not reporting it on the repo as information is so sparse that I cannot make a good bug out of it but wanted to give you a heads up that there might be something wrong "somewhere".
```
_openmp_mutex             Finished _openmp_mutex                        (00m:00s)             435 KB    835 B/s
supervisor                Finished supervisor                           (00m:00s)             547 KB    686 KB/s
libgomp                   Finished libgomp                              (00m:00s)             816 KB    922 KB/s
gettext                   Finished gettext
```

Sometimes things continue but `mamba` output is broken, the following output from `conda` looks fine again:

```
  - sortedcontainers-2.1.0-py_0
  - sphinx-3.0.3-py_0
  - sphinx_rtd_theme-0.4.3-py_0
  - sphinxcontrib-apidoc-0.3.0-py_1
  - sphinxcontrib-applehelp-1.0.2-py_0
  - sphinxcontrib-devhelp-1.0.2-py_0
  - sphinxcontrib-htmlhelp-1.0.3-py_0
  - sphinxcontrib-jsmath-1.0.1-py_0
  - sphinxcontrib-qthelp-1.0.3-py_0
  - sphinxcontrib-serializinghtml-1.1.4-py_0
  - sqlalchemy-1.3.16-py38h1e0a361_0
  - sqlalchemy-utils-0.36.5-pyh9f0ad1d_0
  - sqlite-3
Preparing transaction: ...working... done
Verifying transaction: ...working... done
```

Things I see that could relate to this:

 * Only happens in large environments (>100 packages)
 * Happens more often inside Github Actions than in my local Docker setup.

**Comment by wolfv** [comment]
I assume you're working with the latest release from yesterday night?

There were two bugs that were fixed in relation to downloading: 

- when retrying the previous file wasn't deleted, so that would later fail in validation because it just appended to a previously half-written file
- Validation was crashing when no SHA256 sum existed in the repodata (this seems to be the case for public channels on Anaconda.org, e.g. RoboStack)

But at least with the second issue I mentioned it would fail reliably.

I have also seen timeouts quite frequently with large downloads. Mamba is supposed to retry after a timeout, up to three times, with a backoff factor. Maybe we need to increase the number of retries as well.

Let me add a setting to control how many packages can be downloaded in parallel and then we could try to set that to a more conservative value (e.g. 1 or 2) and see if we observe the same number of errors?

**Comment by wolfv** [comment]
For the timeout you'd see some information on the progress bar, but that information might not appear in the logs. So let me also make sure that that information would appear in the logs.

**Comment by wolfv** [comment]
another thing that would help is if you run with verbose-mode (-v, -vv, and -vvv) although I would definitely only recommend setting `-v` otherwise it will be too much info from libsolv and CURL.

**Comment by xhochy** [comment]
This happens already with 0.2.6 and still in 0.2.8.

I will add some `v`s to the CI.

**Comment by wolfv** [comment]
Thanks! Just for reference this is on Linux?
I feel it could also be related to downloading packages with a super fast internet connection ... i should probably simulate that with a localhost server.

**Comment by xhochy** [comment]
It is on Linux but also happens in Docker-for-Mac with my slower connection (still 100Mbit but not superfast).

**Comment by wolfv** [comment]
ok. Docker-for-Mac I might be able to test over here, or docker in general. Hopefully we can fix this soon.

**Comment by xhochy** [comment]
Related note, we see similar issues with `conda` only on Travis only on `ppc64le` systems: https://github.com/conda-forge/arrow-cpp-feedstock/runs/684238048 Not sure whether you still call any `conda` code in that stage.

**Comment by wolfv** [comment]
hmm. weird! 
your first output looked incomplete to me (the `(00:0X)` which would be printed from the C++ / mamba part. Anyways, I'll add more logging messages so we can see what's going on.

**Comment by xhochy** [comment]
I haven't seen this in a while with mamba 0.2.10+, will reopen if it reappears.

**Comment by wolfv** [comment]
awesome, thanks for letting us know.

**Comment by xhochy** [comment]
Happening again today with `mamba=0.2.12` on CI and locally.

Snippet from the log:

```
INFO    transaction.cpp:151 Adding six with https://conda.anaconda.org/conda-forge/noarch/six-1.15.0-pyh9f0ad1d_0.tar.bz2
INFO    transaction.cpp:151 Adding smmap with https://conda.anaconda.org/conda-forge/noarch/smmap-3.0.4-pyh9f0ad1d_0.tar.bz2
INFO    transaction.cpp:151 Adding snappy with https://conda.anaconda.org/conda-forge/linux-64/snappy-1.1.8-he1b5a44_1.tar.bz2
* Found bundle for host conda.anaconda.org: 0x5634cb810cb0 [serially]
* Found bundle for host conda.anaconda.org: 0x5634cb810cb0 [serially]
* Found bundle for host conda.anaconda.org: 0x5634cb810cb0 [serially]
* Found bundle for host conda.anaconda.org: 0x5634cb810cb0 [serially]
* Found bundle for host conda.anaconda.org: 0x5634cb810cb0 [serially]
* No connections available in cache
* No connections available.
* Found bundle for host conda.anaconda.org: 0x5634cb810cb0 [serially]
* No connections available in cache
* No connections available.
* Found bundle for host conda.anaconda.org: 0x5634cb810cb0 [serially]
* No connections available in cache
* No connections available.
```

The last three lines are repeated then over and over again.

**Comment by wolfv** [comment]
very interesting, thanks for posting the log.
we might have an issue in our logic for parallel downloads then.

Did you see any SHA256 or similar validation errors by chance?

**Comment by dzelge** [comment]
See #301

**Comment by wolfv** [comment]
@xhochy I think this may be the same issue as #301 and hopefully it's fixed by #330! 

I'll make a release right after it's merged.

**Comment by wolfv** [comment]
I am pretty sure that's fixed with #330. It would make a lot of sense.
